File: python_code/ValidSudoku.py
class ValidSudoku:
    @staticmethod
    def is_valid_sudoku(board):
        rows = [[False] * 9 for _ in range(9)]   # ②
        cols = [[False] * 9 for _ in range(9)]   # ③
        boxes = [[False] * 9 for _ in range(9)]  # ④

        for r in range(9):                       # ⑤
            for c in range(9):                   # ⑥
                ch = board[r][c]                 # ⑦
                if ch == '.':                    # ⑧
                    continue

                idx = int(ch) - 1                # ⑨
                box = (r // 3) * 3 + c // 3       # ⑩

                if rows[r][idx] or cols[c][idx] or boxes[box][idx]:  # ⑪
                    return False                 # ⑫

                rows[r][idx] = True              # ⑬
                cols[c][idx] = True              # ⑭
                boxes[box][idx] = True           # ⑮

        return True

# Ejemplo de uso:
board = [
    ['5','3','.','.','7','.','.','.','.'],
    ['6','.','.','1','9','5','.','.','.'],
    ['.','9','8','.','.','.','.','6','.'],
    ['8','.','.','.','6','.','.','.','3'],
    ['4','.','.','8','.','3','.','.','1'],
    ['7','.','.','.','2','.','.','.','6'],
    ['.','6','.','.','.','.','2','8','.'],
    ['.','.','.','4','1','9','.','.','5'],
    ['.','.','.','.','8','.','.','7','9']
]

result = ValidSudoku.is_valid_sudoku(board)
print("¿Tablero válido?:", result)  # Resultado esperado: True

File: java_code/ValidSudoku.java
package java_code;

public class ValidSudoku {
    public static boolean isValidSudoku(char[][] board) {
        boolean[][] rows = new boolean[9][9]; // ②
        boolean[][] cols = new boolean[9][9]; // ③
        boolean[][] boxes = new boolean[9][9]; // ④

        for (int r = 0; r < 9; r++) { // ⑤
            for (int c = 0; c < 9; c++) { // ⑥
                char ch = board[r][c]; // ⑦
                if (ch == '.')
                    continue; // ⑧

                int idx = ch - '1'; // ⑨
                int box = (r / 3) * 3 + c / 3; // ⑩

                if (rows[r][idx] || cols[c][idx] || boxes[box][idx])// ⑪
                    return false; // ⑫

                rows[r][idx] = true; // ⑬
                cols[c][idx] = true; // ⑭
                boxes[box][idx] = true; // ⑮
            }
        }
        return true;
    }

    public static void main(String[] args) {
        char[][] board = {
                { '5', '3', '.', '.', '7', '.', '.', '.', '.' },
                { '6', '.', '.', '1', '9', '5', '.', '.', '.' },
                { '.', '9', '8', '.', '.', '.', '.', '6', '.' },
                { '8', '.', '.', '.', '6', '.', '.', '.', '3' },
                { '4', '.', '.', '8', '.', '3', '.', '.', '1' },
                { '7', '.', '.', '.', '2', '.', '.', '.', '6' },
                { '.', '6', '.', '.', '.', '.', '2', '8', '.' },
                { '.', '.', '.', '4', '1', '9', '.', '.', '5' },
                { '.', '.', '.', '.', '8', '.', '.', '7', '9' }
        };

        isValidSudoku(board);
    }
}
